File: static/js/audit.js
/* Audit Log tab — paginated, type filter, search + date filter. §12 */
(function () {
  "use strict";
  const { api, escapeHtml, debounce } = window.U;

  let offset = 0;
  const LIMIT = 50;

  async function render(container) {
    container.innerHTML = `
      <div class="neo-card filter-bar">
        <div class="filter-row">
          <input class="neo-input" id="aud-q" placeholder="Search description…">
          <select class="neo-input" id="aud-type"><option value="">All event types</option></select>
          <input class="neo-input" id="aud-from" type="date">
          <input class="neo-input" id="aud-to" type="date">
        </div>
        <div class="filter-summary" id="aud-summary"></div>
      </div>
      <div id="aud-list"></div>
      <div class="pager" id="aud-pager" style="display:flex;justify-content:center;gap:12px;padding:18px 0"></div>`;

    const reload = debounce(() => { offset = 0; load(); }, 250);
    ["aud-q","aud-type","aud-from","aud-to"].forEach(id => {
      document.getElementById(id).addEventListener("input", reload);
      document.getElementById(id).addEventListener("change", reload);
    });

    offset = 0;
    await load();
  }

  async function load() {
    const params = new URLSearchParams({
      limit: LIMIT, offset,
      q: document.getElementById("aud-q").value,
      type: document.getElementById("aud-type").value,
      from: document.getElementById("aud-from").value,
      to: document.getElementById("aud-to").value,
    });
    const r = await api(`/api/audit-log?${params}`);
    if (!r.ok) {
      document.getElementById("aud-list").innerHTML = `<div class="neo-card empty-state">Failed to load.</div>`;
      return;
    }

    // Refresh the type dropdown but preserve selection.
    const typeSel = document.getElementById("aud-type");
    const selected = typeSel.value;
    typeSel.innerHTML = `<option value="">All event types (${r.data.total})</option>` +
      r.data.types.map(t => `<option value="${t.event_type}" ${t.event_type === selected ? "selected" : ""}>${escapeHtml(t.event_type)} (${t.count})</option>`).join("");

    document.getElementById("aud-summary").textContent =
      `${r.data.total} events • showing ${r.data.events.length} starting at #${offset + 1}`;

    if (!r.data.events.length) {
      document.getElementById("aud-list").innerHTML =
        `<div class="neo-card empty-state">No events match your filters.</div>`;
    } else {
      document.getElementById("aud-list").innerHTML = `
        <div class="neo-card">
          <table class="audit-table" style="width:100%;border-collapse:collapse">
            <thead><tr><th>When</th><th>Type</th><th>Description</th><th>Linked</th></tr></thead>
            <tbody>
              ${r.data.events.map(e => `<tr>
                <td class="muted">${escapeHtml(e.created_at)}</td>
                <td><span class="badge">${escapeHtml(e.event_type)}</span></td>
                <td>${escapeHtml(e.description)}</td>
                <td class="muted">${e.related_type ? `${escapeHtml(e.related_type)} #${e.related_id ?? "—"}` : "—"}</td>
              </tr>`).join("")}
            </tbody>
          </table>
        </div>`;
    }

    const pager = document.getElementById("aud-pager");
    const hasNext = offset + LIMIT < r.data.total;
    pager.innerHTML = `
      <button class="neo-btn" id="aud-prev" ${offset === 0 ? "disabled" : ""}>← Prev</button>
      <span class="muted" style="align-self:center">page ${Math.floor(offset / LIMIT) + 1}</span>
      <button class="neo-btn" id="aud-next" ${!hasNext ? "disabled" : ""}>Next →</button>`;
    document.getElementById("aud-prev").addEventListener("click", () => { offset = Math.max(0, offset - LIMIT); load(); });
    document.getElementById("aud-next").addEventListener("click", () => { offset += LIMIT; load(); });
  }

  window.AuditLog = { render };
})();

File: routes/audit.py
"""Read-only audit log endpoint (§12). Paginated, newest first, filter by type."""
from flask import Blueprint, jsonify, request

from database import get_db
from routes.auth import login_required

bp = Blueprint("audit", __name__, url_prefix="/api/audit-log")


@bp.get("")
@login_required
def list_audit_log():
    db = get_db()
    args = request.args
    where = []
    params = []
    if t := args.get("type"):
        where.append("event_type = ?")
        params.append(t)
    if q := args.get("q"):
        where.append("description LIKE ?")
        params.append(f"%{q}%")
    if df := args.get("from"):
        where.append("created_at >= ?")
        params.append(df)
    if dt := args.get("to"):
        where.append("created_at <= ?")
        params.append(dt + " 23:59:59")

    try:
        limit = min(max(1, int(args.get("limit", 50))), 500)
        offset = max(0, int(args.get("offset", 0)))
    except ValueError:
        limit, offset = 50, 0

    where_sql = " AND ".join(where) if where else "1=1"
    rows = db.execute(
        f"SELECT id, event_type, description, related_type, related_id, created_at "
        f"FROM audit_log WHERE {where_sql} ORDER BY id DESC LIMIT ? OFFSET ?",
        (*params, limit, offset),
    ).fetchall()
    total = db.execute(
        f"SELECT COUNT(*) AS c FROM audit_log WHERE {where_sql}", params
    ).fetchone()["c"]
    types = db.execute(
        "SELECT event_type, COUNT(*) AS c FROM audit_log GROUP BY event_type ORDER BY event_type"
    ).fetchall()
    return jsonify(
        events=[dict(r) for r in rows],
        total=total,
        limit=limit,
        offset=offset,
        types=[{"event_type": t["event_type"], "count": t["c"]} for t in types],
    )
